packages/flame/lib/src/game/game_widget/game_widget.dart

```diff
@@ -151,55 +142,61 @@ class _GameWidgetState<T extends Game> extends State<GameWidget<T>> {
 
   late FocusNode _focusNode;
 
-  @override
-  void initState() {
-    super.initState();
-
-    // Add the initial overlays
-    _initActiveOverlays();
-    addOverlaysListener();
+  /// The number of `build()` functions currently executing.
+  int _buildDepth = 0;
 
-    // Add the initial mouse cursor
-    _initMouseCursor();
-    addMouseCursorListener();
+  /// If true, then a fresh build will be scheduled after the current one
+  /// completes. This should only be set to true when the [_buildDepth] is
+  /// non-zero.
+  bool _requiresRebuild = false;
 
-    _focusNode = widget.focusNode ?? FocusNode();
-    if (widget.autofocus) {
-      _focusNode.requestFocus();
+  /// Helper method that arranges to have `_buildDepth > 0` while the [build] is
+  /// executing, and then schedules a re-build if [_requiresRebuild] flag was
+  /// raised during the build.
+  ///
+  /// This is needed because our build function invokes user code, which in turn
+  /// may change some of the [Game]'s properties which would require the
+  /// [GameWidget] to be rebuilt. However, Flutter doesn't allow widgets to be
+  /// marked dirty while they are building. So, this method is needed to avoid
+  /// such a limitation and ensure that the user code can set [Game]'s
+  /// properties freely, and that they will be propagated to the [GameWidget]
+  /// at the earliest opportunity.
+  Widget _protectedBuild(Widget Function() build) {
```

Comment:
This function concerns me.

A `Widget.build` method should be pure, so it shouldn't trigger any side effects.

I am not sure I completely understand why it is needed, before we did everything on the `initState` can't we keep doing that?

Reply:
Yeah, I get it, but I don't like the idea of those side effects, maybe we should completely disallow user to do those modifications on build time of the game widget, and provide initial properties for setting the initial values (which is something I believe we already do?). And then just throw meaningful messages when the user tries to change something that they shouldn't, like we do with the `hasLayout` method.

Reply:
I understand that this makes the code logic slightly more complicated and "less pure". So it's a detriment to us, developers. At the same time, from the user's perspective, the ability to set up their `Game` from within the class itself -- is a very welcome one. Setting these same properties from `GameWidget` is not as convenient, since it breaks encapsulation of that logic. 

So, it really comes down as a choice between developer convenience and user convenience.

Another difficulty with "completely disallowing user to do these modifications during build time" is that _build time_ is not a well-defined concept within Flame. If it was just the case that the `GameWidget` is built once and for all, then it might be feasible.  But the widget is constantly rebuilt, either due to changes in system environment, or because of changes in the surrounding widget tree, or because of changes coming internally from the `Game`. So how do we explain all these complexities to the user? Do we even understand them ourselves?

The solution in this PR ensures that the widget works regardless of the external or internal sequence of events.

Reply:
> The solution in this PR ensures that the widget works regardless of the external or internal sequence of events.

It does, but the solution relies on build side effects, which is the problem for me.

I wonder if we couldn't queue the events when the game is not attached yet, and the. Trigger them on the game attachment

Reply:
> I wonder if we couldn't queue the events when the game is not attached yet, and then trigger them on the game attachment

This is kinda what this PR is doing, isn't it? The events are "queued" while the game is being built, and then re-triggered at the end of the build. 
Handling them at the end of attachment wouldn't work because a build can trigger multiple times during the course of the game, and we want those builds to be protected too.

> It does, but the solution relies on build side effects, which is the problem for me.

The only real "side effect" here is that a build can schedule a future rebuild if there were modifications to `Game` instance during that build. I'm not sure how this can be considered a "problem" if the alternatives -- either to do nothing and have the changes not being reflected in the UI, or to let the game crash (the current behavior) -- seem much worse.

Reply:
Any changes to a state from inside the build is a side effect, that is highly discouraged, [this](https://codewithandrea.com/articles/side-effects-flutter/) simplifies the problem but it can give a rough idea.

If the only way to solve the issue is by the way it is implemented now, I would rather keep the user informing the initial mouse cursor and overlays on the game widget and the user can only change those after the game is attached.

But I will spend some time taking a look on this and testing a few ideas.

Reply:
> It seems like this could be a situation were the rules can be bent.

It may be indeed! But I want to explore a little bit more before we go on that path, I will report back here.

Reply:
Okay, I did some research, and it seems that having things like this on more complicated scenarios is not that bad, provider for example does a bunch of stuff in the same idea that were implemented here. So I think I am okay with going with this approach.